api/internals/init.md

```diff
@@ -0,0 +1,169 @@
+# IPFS Node Init Specification
+
+> This is the specification for the steps performed when a new IPFS node is initialized.
+
+*Note that initializing an IPFS node differs from initializing an IPFS repo. For more on repo initialization, read the [Repo specification](https://github.com/ipfs/specs/tree/master/repo).*
+
+## Status
+
+First draft.
+
+For now, [`ipfs/go-ipfs/cmd/ipfs/init.go`](https://github.com/ipfs/go-ipfs/blob/master/cmd/ipfs/init.go) is the source of truth.
+
+## Flags
+
+There are some optional flags that may be supplied to `ipfs init`:
+
+```
+ -f|--force: whether to overwrite an existing repo config. Defaults to false.
+ -e|--empty-repo: whether to add+pin ipfs info help files to the repo. Defaults to false.
+ -b|--bits: # of bits to use in the generated RSA private key. Defaults to 2048.
+```
+
+## Initialization Sequence
+
+### 1. Verify Initial Conditions
+
+Before initialization can begin, the following must be true:
+
+1. The above input flags are parsed and validated.
+2. The [config file](https://github.com/ipfs/specs/tree/master/repo#config) has not been locked by an active daemon process.
+3. The repo root is writeable.
+4. The repo either doesn't exist and hasn't been initialized, or does already exist but the `--force` flag is supplied.
+
+If all of these are satisfied, output `"initializing ipfs node at %s\n", repoRoot` to standard out and proceed.
+
+### 2. Generate Repo Configuration
+
+These are the steps to producing the default [config tree](https://github.com/ipfs/specs/tree/master/repo#config). The actual output format will depend on the repo backend (i.e. likely JSON if fs-repo). Check the specification for the relevant backend if unsure.
+
+The following sections describe the various subtrees in the config and how to generate them. For simplicity and consistency, JSON will be used.
+
+#### Identity
+
+The RSA keypair must be >= 1024 bits.
+
+Before beginning key generation, write `"generating %v-bit RSA keypair...", nbits` to standard out. (no newline)
+
+Generate the keypair. If successful, append `"done\n"` to standard out.
+
+The private key must be formatted as the base64 encoding of its bytes. It is currently stored unencrypted.
+
+The PeerID must be formatted as
+
+```json
+Base58.Encode(Multihash('SHA2-256', PublicKey.Bytes))
+```
+
+Resulting in the following config tree entry:
+
+```json
+"Identity": {
+	"PeerID": "<PeerID>",
+	"PrivKey": "<PrivKey>"
+}
+```
```

Comment:
It was my understanding that one could init a repo with go-ipfs, then launch js-ipfs and point it at the config created by go-ipfs. If this is true -- that configs ought to be compatible across implementations -- we need to have a very consistent format.